### PR TITLE
Prevent flooding (and app freeze) of items added to the message bar

### PR DESCRIPTION
**Pull request by nirvn**
## Description
This PR adds some flooding prevention to the message bar by:
- limiting the number of items contained in the message bar to _20_ (I'm open to the maximum count here, but really, if the bar has more than 20 items, something has gone wrong already :) )
- preventing item duplicates (defined by item's title, text, and level being equal)

@nyalldawson , as discussed on hangout.

This prevents QGIS freezing when flooded with auth error messages:
image

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nirvn** [comment]
Spotted in the wild: https://issues.qgis.org/issues/20262

**Comment by luipir** [comment]
hi @nirvn becasue these parts are not touched since time, and right now we are experiencing a set of errors related with auth system (especially in tile services). can you give us a hint if this service was working before with any 3.x release or you just faced it right now?

**Comment by nirvn** [comment]
@luipir , I never stumbled on these errors until way into the 3.3 dev cycle. It's not straightforward to reproduce reliably, you may have to pan, zoom in and out for a while prior to being flooded.

For the record: the auth error needs addressing, the PR isn't about that (nor about masking the error since the error will show in the bar), it's only a way to prevent UI freeze / spam.

**Comment by luipir** [comment]
@nirvn what platform? I suspect it's more a packaging problem
btw I understand the pr is not related with this error

**Comment by nirvn** [comment]
@luipir , ubuntu 18.10, self-made build.

**Comment by nirvn** [comment]
(moved discussion off this PR)

File: src/gui/qgsmessagebar.cpp [hunk 1]
```diff
@@ -298,6 +298,17 @@ QgsMessageBarItem *QgsMessageBar::pushWidget( QWidget *widget, Qgis::MessageLeve
 
 void QgsMessageBar::pushMessage( const QString &title, const QString &text, Qgis::MessageLevel level, int duration )
 {
+  // refrain from adding more than 20 items in the message bar, avoids flooding (and freezing) of the main window
+  if ( mItems.count() > 20 )
```
Comment (nyalldawson) [inline src/gui/qgsmessagebar.cpp]: Instead of blocking new items in this case, would we be better off removing the earliest one?

Comment (nirvn) [reply src/gui/qgsmessagebar.cpp]: @nyalldawson , I'm not sure. It'd be costlier to do so, and it seems would defeat the flooding protection intent here. Also, for items served with a timeout, it's essentially reset the timeout every time a duplicate item is added (which means that in the case of the auth error, it'd stick there for a looongggg time :) )

Comment (nirvn) [reply src/gui/qgsmessagebar.cpp]: @nyalldawson , gave this some more thoughts, and adjusted the commit to act as per your suggestion.

**Pull request merged**